git-cliff-core/src/repo.rs

```diff
@@ -76,36 +86,104 @@ impl Repository {
 			.collect();
 		if include_path.is_some() || exclude_path.is_some() {
 			commits.retain(|commit| {
-				if let Ok(prev_commit) = commit.parent(0) {
-					if let Ok(diff) = self.inner.diff_tree_to_tree(
-						commit.tree().ok().as_ref(),
-						prev_commit.tree().ok().as_ref(),
-						None,
-					) {
-						return diff
-							.deltas()
-							.filter_map(|delta| delta.new_file().path())
-							.any(|new_file_path| {
-								if let Some(include_path) = &include_path {
-									include_path
-										.iter()
-										.any(|glob| glob.matches_path(new_file_path))
-								} else if let Some(exclude_path) = &exclude_path {
-									!exclude_path
-										.iter()
-										.any(|glob| glob.matches_path(new_file_path))
-								} else {
-									false
-								}
-							});
-					}
-				}
-				false
+				self.should_retain_commit(commit, &include_path, &exclude_path)
 			});
 		}
 		Ok(commits)
 	}
 
+	/// Calculates whether the commit should be retained or not.
+	fn should_retain_commit(
+		&self,
+		commit: &Commit,
+		include_path: &Option<Vec<Pattern>>,
+		exclude_path: &Option<Vec<Pattern>>,
+	) -> bool {
+		let changed_files = self.commit_changed_files(commit);
+
+		if let Some(include_path) = include_path {
+			if !include_path.iter().any(|pattern| {
+				changed_files.iter().any(|path| pattern.matches_path(path))
+			}) {
+				return false;
+			}
+		}
+
+		if let Some(exclude_path) = exclude_path {
+			if exclude_path.iter().any(|pattern| {
+				changed_files.iter().any(|path| pattern.matches_path(path))
+			}) {
+				return false;
+			}
+		}
+
+		true
+	}
+
+	fn commit_changed_files(&self, commit: &Commit) -> Vec<PathBuf> {
+		// Cache key is generated from the repository path and commit id
+		let cache_key = format!("commit_id:{}", commit.id());
+
+		// Check the cache first.
+		{
+			if let Ok(result) =
+				cacache::read_sync(&self.changed_files_cache_path, &cache_key)
+			{
+				if let Ok((files, _)) =
+					bincode::decode_from_slice(&result, bincode::config::standard())
+				{
+					return files;
+				}
+			}
+		}
+
+		// If the cache is not found, calculate the result and set it to the cache.
+		let result = self.commit_changed_files_no_cache(commit);
+
+		// Set the result to the cache.
+		match bincode::encode_to_vec(&result, bincode::config::standard()) {
```

Comment:
Just curious why we need `bincode` here

Reply:
Bincode helps with serializing the paths so that they can be stored in the cache. I tried serde_json first, but the whole generation ended up being 2.5 times slower than using bincode. So, I chose bincode to avoid adding too much caching overhead.